validator/accounts/status.go

```diff
@@ -0,0 +1,176 @@
+package accounts
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+	"sort"
+
+	"github.com/pkg/errors"
+	ethpb "github.com/prysmaticlabs/ethereumapis/eth/v1alpha1"
+	"github.com/prysmaticlabs/prysm/shared/keystore"
+	"go.opencensus.io/trace"
+)
+
+// ValidatorStatusMetadata holds all status information about a validator.
+type ValidatorStatusMetadata struct {
+	PublicKey []byte
+	Metadata  *ethpb.ValidatorStatusResponse
+}
+
+// MaxRequestLimit specifies the max grpc requests allowed
+// to fetch account statuses.
+const MaxRequestLimit = 3
+
+// MaxRequestKeys specifies the max amount of public keys allowed
+// in a single grpc request, when fetching account statuses.
+const MaxRequestKeys = 2000 // XXX: This is an arbitrary number.
+// Should compute max keys allowed before exceeding GrpcMaxCallRecvMsgSizeFlag
+
+// FetchAccountStatuses fetches validator statuses from the BeaconNodeValidatorClient
+// for each validator public key.
+func FetchAccountStatuses(
+	ctx context.Context,
+	beaconNodeRPCProvider ethpb.BeaconNodeValidatorClient,
+	pubkeys [][]byte) ([][]ValidatorStatusMetadata, error) {
+	ctx, span := trace.StartSpan(ctx, "validator.FetchAccountStatuses")
+	defer span.End()
+
+	errorChannel := make(chan error, MaxRequestLimit)
+	statusChannel := make(chan []ValidatorStatusMetadata, MaxRequestLimit)
+	// Launch routines to fetch statuses.
+	i, numBatches := 0, 0
+	for ; i+MaxRequestKeys < len(pubkeys); i += MaxRequestKeys {
+		go fetchValidatorStatus(
+			ctx, beaconNodeRPCProvider, pubkeys[i:i+MaxRequestKeys], statusChannel, errorChannel)
+		numBatches++
+	}
+	if i < len(pubkeys) {
+		go fetchValidatorStatus(
+			ctx, beaconNodeRPCProvider, pubkeys[i:], statusChannel, errorChannel)
+		numBatches++
+	}
+	// Wait for fetch routines to finish.
+	var err error
+	allStatuses := make([][]ValidatorStatusMetadata, 0, numBatches)
+	for i := 0; i < numBatches; i++ {
+		select {
+		case statuses := <-statusChannel:
+			allStatuses = append(allStatuses, statuses)
+		case e := <-errorChannel:
+			err = e
+		}
+	}
+
+	return allStatuses, err
+}
+
+func fetchValidatorStatus(
+	ctx context.Context,
+	rpcProvder ethpb.BeaconNodeValidatorClient,
+	pubkeys [][]byte,
+	statusChannel chan []ValidatorStatusMetadata,
+	errorChannel chan error) {
+	if ctx.Err() == context.Canceled {
+		errorChannel <- errors.Wrap(ctx.Err(), "context has been canceled.")
+		return
+	}
+
+	req := &ethpb.MultipleValidatorStatusRequest{PublicKeys: pubkeys}
+	resp, err := rpcProvder.MultipleValidatorStatus(ctx, req)
+	if err != nil {
+		errorChannel <- errors.Wrap(
+			err,
+			fmt.Sprintf("Failed to fetch validator statuses for %d key(s)", len(pubkeys)))
+		return
+	}
+
+	statusChannel <- responseToSortedMetadata(resp)
+}
+
+func responseToSortedMetadata(resp *ethpb.MultipleValidatorStatusResponse) []ValidatorStatusMetadata {
```

Comment:
No need for this one-off helper func as its logic is quite small where it is used.

Reply:
https://github.com/prysmaticlabs/prysm/pull/5784/commits/a2422db365676bc8f1315aa44fd4953e7c2212a1